ksql-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenSpec.java

```diff
@@ -137,46 +133,145 @@ CodeGenSpec build() {
+        final Class<?> type
+    ) {
+      this.name = requireNonNull(name, "name");
+      this.type = requireNonNull(type, "type");
     }
 
+    @Override
     public String name() {
       return name;
     }
 
+    @Override
     public Class<?> type() {
       return type;
     }
+  }
 
-    public OptionalInt colIndex() {
-      return columnIndex;
+  @Immutable
+  private static final class FunctionArgumentSpec extends BaseArgumentSpec {
+
+    @EffectivelyImmutable
+    private final Kudf kudf;
+
+    FunctionArgumentSpec(
+        final String name,
+        final Class<?> type,
+        final Kudf kudf
+    ) {
+      super(name, type);
+      this.kudf = requireNonNull(kudf, "kudf");
     }
 
-    public Optional<Kudf> kudf() {
+    @Override
+    public Object resolve(final Object key, final GenericRow value) {
       return kudf;
     }
 
     @Override
     public String toString() {
-      return "ArgumentSpec{"
-          + "name='" + name + '\''
-          + ", type=" + type
-          + ", columnIndex=" + columnIndex
+      return "FunctionArgumentSpec{"
+          + "name='" + name() + '\''
+          + ", type=" + type()
           + ", kudf=" + kudf
           + '}';
     }
   }
 
+  @Immutable
+  public static final class ValueArgumentSpec extends BaseArgumentSpec {
+
+    private final int columnIndex;
+
+    ValueArgumentSpec(
+        final String name,
+        final Class<?> type,
+        final int columnIndex
+    ) {
+      super(name, type);
+      this.columnIndex = columnIndex;
+    }
+
+    @Override
+    public Object resolve(final Object key, final GenericRow value) {
+      return value.getColumns().get(columnIndex);
+    }
+
+    @Override
+    public String toString() {
+      return "ValueArgumentSpec{"
+          + "name='" + name() + '\''
+          + ", type=" + type()
+          + ", columnIndex=" + columnIndex
+          + '}';
+    }
+  }
+
+  @Immutable
+  public static final class KeyArgumentSpec extends BaseArgumentSpec {
+
+    private final int columnIndex;
+
+    KeyArgumentSpec(
+        final String name,
+        final Class<?> type,
+        final int columnIndex
+    ) {
+      super(name, type);
+      this.columnIndex = columnIndex;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Object resolve(final Object key, final GenericRow value) {
+      if (key == null) {
+        return null;
+      }
+
+      return key instanceof Windowed
+          ? resolveWindowed((Windowed<Struct>)key)
+          : resolveNonWindowed((Struct)key);
+    }
+
+    private Object resolveWindowed(final Windowed<Struct> windowedKey) {
+      final Window window = windowedKey.window();
+      final long start = window.start();
+      final String end = window instanceof SessionWindow ? String.valueOf(window.end()) : "-";
+      final Object key = resolveNonWindowed(windowedKey.key());
+      return String.format("%s : Window{start=%d end=%s}", key, start, end);
```

Comment:
This is a copy of what is currently done in `SourceBuilder`.   This is how it handles windowed keys. Yuck!  

Copy in `SourceBuilder` will go soon.

Will be fixed in time by structured keys work.